adage/analyze/api.py

```diff
@@ -472,8 +472,30 @@ class Meta:
             'signature': ('exact', 'in', ),
             'gene': ('exact', 'in', ),
             'participation_type': ('exact', 'in', ),
+            'related_genes': ('exact')  # see apply_filters()
         }
 
+    def apply_filters(self, request, applicable_filters):
+        """Implementation of "related_genes" filter, which allows the
+        API to get all participation records of signatures that include
+        at least one of the entries in "related_genes" list.
+        This filter is used to calculate enriched signatures given a
+        list of genes.
+        """
+        object_list = super(ParticipationResource, self).apply_filters(
+            request, applicable_filters)
+        related_genes = request.GET.get('related_genes', None)
+        if related_genes:
+            try:
+                query_genes = {int(id) for id in related_genes.split(',')}
+            except ValueError:
+                raise BadRequest("Invalid gene IDs: %s" % related_genes)
+
+            signatures = Participation.objects.filter(
+                gene__in=query_genes).values('signature').distinct()
+            object_list = object_list.filter(signature__in=signatures)
```

Comment:
I tested the query above and it was slow. My first thought was to create indexes for `signature` and `gene` fields in `Participation` table, but then I realized that both are foreign keys, so they are already indexed. Due to the nature of this query, I don't think it can be done by going through the participation records just once. One thing I can think of is to get all participation records first from the database, then filter the records based on input `related_genes`.

Reply:
@mhuyck: I did a few more tests and found that the most time consuming part is actually this line:
```
object_list = object_list.filter(signature__in=signatures)
```
because `gene` field is a foreign key in `ParticipationResource`, the more records in the returned `object_list`, the more table joins will be needed to get the details of gene. Since `signature` and the current module do need the gene information, we can not change line 461 into:
```
gene = fields.IntegerField(attribute='gene_id', null=False)
```

Reply:
Rats. I thought this would be simpler to speed up. If it's not a quick thing to optimize the filter then we should log it as an issue and come back to it. It's better to *have* the feature and see how it gets used before spending a lot of effort to make it perfect.

Reply:
☝️  Do we have a spinner we could easily have up as a filler while this happens so the user knows something is being done?

Reply:
Let me find a spinner.